specifyweb/businessrules/views.py

```diff
@@ -1 +1,288 @@
+                            }
+                        }
+                    }
+                }
+            }
+        },
+        "responses": {
+            "201": {
+                "description": "Uniqueness rules properly updated and/or created",
+                "content": {
+                    "application/json": {
+                        "schema": {
+                            "type": "object"
+                        }
+                    }
+                }
+            }
+        }
+    }
+})
+@login_maybe_required
+@require_http_methods(['GET', 'PUT', 'POST'])
+@transaction.atomic
+def uniqueness_rule(request, discipline_id):
+    data = {}
+
+    try:
+        model = request.GET["model"]
+    except:
+        model = None
+
+    if request.method == 'GET':
+        rules = UniquenessRule.objects.filter(discipline=discipline_id)
+        for rule in rules:
+            rule_fields = rule.splocalecontaineritems.filter(
+                uniquenessrule_splocalecontaineritem__isScope=0)
+            scope = rule.splocalecontaineritems.filter(
+                uniquenessrule_splocalecontaineritem__isScope=1)
+
+            table = rule_fields[0].container.name
+            if model is not None and table.lower() != model.lower():
+                continue
+            if table not in data.keys():
+                data[table] = []
+            data[table].append({"rule": {"id": rule.id, "fields": [{"id": field.id, "name": field.name} for field in rule_fields], "scopes": [{
+                               "id": _scope.id, "name": _scope.name} for _scope in scope], "isDatabaseConstraint": rule.isDatabaseConstraint}})
+
+    elif request.method == 'PUT' or request.method == 'POST':
+        rules = json.loads(request.body)['rules']
+        discipline = models.Discipline.objects.get(id=discipline_id)
+        for rule in rules:
+            fetched_scopes = models.Splocalecontaineritem.objects.filter(
+                id__in=[int(scope["id"]) for scope in rule["scopes"]]) if len(rule["scopes"]) > 0 else [None]
+            if rule["id"] is None:
+                fetched_rule = UniquenessRule.objects.create(
+                    isDatabaseConstraint=rule["isDatabaseConstraint"], discipline=discipline)
+            else:
+                fetched_rule = UniquenessRule.objects.get(id=rule["id"])
+                fetched_rule.discipline = discipline
+                fetched_rule.isDatabaseConstraint = rule["isDatabaseConstraint"]
+                fetched_rule.save()
+
+            fetched_rule.splocalecontaineritems.clear()
+            locale_items = models.Splocalecontaineritem.objects.filter(
+                id__in=[field["id"] for field in rule["fields"]])
+
+            fetched_rule.splocalecontaineritems.set(list(locale_items))
+            fetched_rule.splocalecontaineritems.add(
+                *fetched_scopes, through_defaults={"isScope": True})
+
+            make_uniqueness_rule(fetched_rule)
```

Comment:
I have since implemented a system which disconnects previously connected signals. 
I utilize this via Django's [dispatch_uid](https://docs.djangoproject.com/en/5.0/topics/signals/#preventing-duplicate-signals) for signals to hold references to them. 
Thus when the rules are updated the dispatch_uid is used to disconnect the previous rule and re-establish the rule. 

Would this be sufficient? Or do you think the ideas you proposed above be more readable and/or less complex?

Reply:
Love the current solution! 

Just a nitpick: can you prepend `uniqueness-rules-` to the rule ids? So, it becomes `uniqueness-rules-1` ... `uniqueness-rules-2`. The intention here is that if a similar solution (disconnecting via uid) comes up, it will not conflict.